Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -0,0 +1,504 @@
+    /// When the pool is asked for a new connection, it creates a `Key` from the url associated to the `request`. This key
+    /// is used to determine if there already exists an associated `ConnectionProvider` in `connectionProviders`
+    /// if it does, the connection provider then takes care of leasing a new connection. If a connection provider doesn't exist, it is created.
+    func getConnection(for request: HTTPClient.Request, preference: HTTPClient.EventLoopPreference, deadline: NIODeadline?) -> EventLoopFuture<Connection> {
+        let key = Key(request: request)
+        // FIXME: Is this correct to use self.loopGroup?
+        switch self.getAction(for: key, eventLoop: self.loopGroup.next()) {
+        case .useExisting(let provider):
+            return provider.getConnection(preference: preference)
+
+        case .make(completing: let providerPromise):
+            // If no connection provider exists for the given key, create a new one
+            let selectedEventLoop = providerPromise.futureResult.eventLoop
+            var bootstrap = ClientBootstrap.makeHTTPClientBootstrapBase(group: selectedEventLoop, host: request.host, port: request.port, configuration: self.configuration) { _ in
+                self.loopGroup.next().makeSucceededFuture(())
+            }
+
+            if let timeout = resolve(timeout: self.configuration.timeout.connect, deadline: deadline) {
+                bootstrap = bootstrap.connectTimeout(timeout)
+            }
+
+            let address = HTTPClient.resolveAddress(host: request.host, port: request.port, proxy: self.configuration.proxy)
+
+            return bootstrap.connect(host: address.host, port: address.port).flatMap { channel in
+                let http1PipelineConfigurator: (ChannelPipeline) -> EventLoopFuture<ConnectionProvider> = { pipeline in
+                    pipeline.addHTTPClientHandlers(leftOverBytesStrategy: .forwardBytes).map { _ in
+                        let http1Provider = ConnectionProvider.http1(HTTP1ConnectionProvider(group: self.loopGroup, key: key, configuration: self.configuration, initialConnection: Connection(key: key, channel: channel, parentPool: self), parentPool: self))
+                        return http1Provider
+                    }
+                }
+
+                return channel.pipeline.addSSLHandlerIfNeeded(for: key, tlsConfiguration: self.configuration.tlsConfiguration).flatMap { _ in
+                    http1PipelineConfigurator(channel.pipeline)
+                }.flatMap { provider in
+                    providerPromise.succeed(provider)
+                    return provider.getConnection(preference: preference)
+                }
+            }
+        }
+    }
+
+    func release(_ connection: Connection) {
+        self.lock.withLock {
+            if let provider = self._connectionProviders[connection.key] {
+                provider.release(connection: connection)
+            }
+        }
+    }
+
+    func syncClose() throws {
```

Comment:
Should this put us into some kind of unrecoverable state? Right now you can `close` this but it'll keep working.

Reply:
Are you thinking to something such as adding a `isClosed` property to `State` and checking that this property isn't set to `true` on all calls to the `HTTP1ConnectionProvider`? Should this be checked in debug only?

Related: The https://github.com/swift-server/async-http-client/pull/105/commits/7b258f9919cb74715238ce847ffa835b46edb1c2 commit adds additional checks on close.

Reply:
Yeah, something like that. And it should be checked in release too: if you closed a connection pool, you shouldn't ever be able to use it again.